### PR TITLE
feat: azure arc addon

### PR DESCRIPTION
**Pull request by jadarsie**
**Reason for Change**:
New addon to simplify cluster onboarding to Azure Arc for Kubernetes.

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jadarsie** [comment]
Labeled as WIP until the onboarding image is hosted in MCR.

**Comment by jadarsie** [comment]
> Labeled as WIP until the onboarding image is hosted in MCR.

Apparently this is not a requirement any more. Removing label.

**Comment by jadarsie** [comment]
> Labeled as WIP until the onboarding image is hosted in MCR.

Apparently this is not a requirement any more. Removing label.

File: pkg/engine/engine.go [hunk 1]
```diff
@@ -886,6 +886,32 @@ func getClusterAutoscalerAddonFuncMap(addon api.KubernetesAddon, cs *api.Contain
 	}
 }
 
+func getArcAddonFuncMap(addon api.KubernetesAddon) template.FuncMap {
```
Comment (jackfrancis) [inline pkg/engine/engine.go:889]: I don't think we need this template funcmap, we can use `getAddonFuncMap`, which already has a way to get the config key vals (`ContainerConfig`)

File: docs/topics/clusterdefinitions.md [hunk 2]
```diff
@@ -128,6 +128,7 @@ $ aks-engine get-versions
...
 | [csi-secrets-store](../../examples/addons/csi-secrets-store/README.md)                                | true (for 1.16+ clusters)                                                                                | as many as linux agent nodes    | Integrates secrets stores (Azure keyvault) via a [Container Storage Interface (CSI)](https://kubernetes-csi.github.io/docs/) volume.                                                                                                                                                                    |
+| [arc](../../examples/addons/arc/README.md)                                  | false                                                                                | 7                               | Attaches the cluster to Azure Arc for Kubernetes. |
```
Comment (jackfrancis) [inline docs/topics/clusterdefinitions.md:131]: slight preference for calling this addon `azure-arc-onboarding` so it's clear that it only delivers that "container" in a job.

**Comment by jackfrancis** [comment]
We'll want to add a default addon data model via `pkg/api/addons.go` (hopefully the existing addons will provide ample examples)

File: pkg/engine/engine.go (hunk 1 above)
Comment (jadarsie) [reply pkg/engine/engine.go]: Added `ContainerConfigBase64` func

File: examples/addons/azure-arc-onboarding/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+| tenantID         | yes      | Tenant ID that owns the specified Subscription                                         |
+| resourceGroup    | yes      | Existing resource group name where the `connectedCluster` ARM resource will be created |
+| clusterName      | yes      | Unique cluster friendly name                                                           |
+| clientID         | yes      | Service principal ID with permissions to create resources in target subscription/group |
+| clientSecret     | yes      | Service principal secret                                                               |
+
+Example:
+
+```json
+{
```
Comment (jackfrancis) [inline examples/addons/azure-arc-onboarding/README.md:29]: let's enclose this example JSON in something like this:

```json
{
  ...
  "properties": {
    "orchestratorProfile": {
      "kubernetesConfig": {
        "addons": [
          ...,
          <your example addons configuration here>,
          ...
        ]
      }
    },
    ...
}
```

**Comment by jackfrancis** [comment]
lgtm, thanks @jadarsie!

File: test/e2e/runner/cli_provisioner.go [hunk 4]
```diff
@@ -444,11 +444,9 @@ func (cli *CLIProvisioner) FetchActivityLog(acct *azure.Account, logPath string)
 func (cli *CLIProvisioner) EnsureArcResourceGroup() error {
 	for _, addon := range cli.Engine.ClusterDefinition.Properties.OrchestratorProfile.KubernetesConfig.Addons {
 		if addon.Name == common.AzureArcOnboardingAddonName && to.Bool(addon.Enabled) {
-			rg := cli.Account.ResourceGroup
-			if err := cli.Account.CreateGroupWithRetry(addon.Config["resourceGroup"], addon.Config["location"], 30*time.Second, cli.Config.Timeout); err != nil {
-				return errors.Wrapf(err, "Error while trying to create Azure Arc resource group: %s", rg)
+			if err := cli.Account.CreateArcGroupWithRetry(addon.Config["resourceGroup"], addon.Config["location"], 30*time.Second, cli.Config.Timeout); err != nil {
```
Comment (jackfrancis) [inline test/e2e/runner/cli_provisioner.go:447]: @jadarsie why do we need to create a new func for this?

File: test/e2e/runner/cli_provisioner.go [hunk 5]
```diff
@@ -205,6 +205,8 @@ func (cli *CLIProvisioner) provision() error {
 	}
 	cli.Engine = eng
 
+	cli.EnsureArcResourceGroup()
```
Comment (jackfrancis) [inline test/e2e/runner/cli_provisioner.go:208]: O.K., lets add this after L30 above (after we've called `CreateGroupWithRetry` and received a nil err):

```
now := fmt.Sprintf("now=%v", time.Now().Unix())
cli.Account.ResourceGroup = r := ResourceGroup{
	Name:     cli.Config.Name,
	Location: cli.Config.Location,
	Tags: map[string]string{
		"now": now,
	},
}
```

Then we can get rid of the `now` var in the `CreateGroup` method, and get rid of the whole "assign ResourceGroup property to account instance" code.

I think that's the easiest path forward. You can then get rid of the new duplicated "create resource group func" and basically do exactly what you were doing before, without the surprise Account.ResourceGroup override. :)

**Pull request merged**